python/test/unit/language/test_conversions.py

```diff
@@ -0,0 +1,308 @@
+
+    idxs = tl.program_id(0) * BLOCK_SIZE + tl.arange(0, BLOCK_SIZE)
+
+    x = tl.load(src + idxs)
+
+    if numbits_src == 8:
+        x = x.to(tl.uint8, bitcast=True)
+    elif numbits_src == 16:
+        x = x.to(tl.uint16, bitcast=True)
+
+    x = x.to(tl.uint32)
+
+    mantissa_mask : tl.constexpr = (1 << mantissa_bits) - 1
+    exponent_mask : tl.constexpr = (1 << exponent_bits) - 1
+
+    mantissa = x & mantissa_mask
+    exponent = (x >> mantissa_bits) & exponent_mask
+    sign = (x >> (numbits_src - 1))
+
+    y = (sign << 31) | (exponent << 23) | (mantissa << (23 - mantissa_bits))
+    y = y.to(tl.float32, bitcast=True)
+    y = y * exponent_compensator
+
+    tl.store(dst + idxs, y)
+
+
+def launch_upcast_emulated(src, exponent_bits, mantissa_bits, exponent_bias, BLOCK_SIZE=4096):
+
+    dst = torch.empty(src.shape, dtype=torch.float32, device='cuda')
+    upcast_emulated[(src.shape[0] // BLOCK_SIZE,)](src, dst, BLOCK_SIZE, exponent_bits, mantissa_bits, exponent_bias)
+    return dst
+
+
+def sizeof(t):
+
+    if t == torch.float64:
+        return 8
+    elif t == torch.float32:
+        return 4
+    elif t == torch.float16 or t == torch.bfloat16:
+        return 2
+    else:
+        return 1
+
+
+def downcast_test(src_dtype, dst_dtype, rounding, exponent_bits, mantissa_bits, exponent_bias, max_repr, offset):
+
+    src = launch_exhaustive_populate(src_dtype, offset << 24, 2**24, False, sizeof(src_dtype) * 8, max_repr)
+    dst = launch_type_convert_triton(src, dst_dtype, rounding)
+    src = launch_type_convert_triton(src, torch.float32)
+
+    dst2 = launch_downcast_emulated(src, dst_dtype, rounding, exponent_bits, mantissa_bits, exponent_bias)
+
+    dst = launch_upcast_emulated(dst, exponent_bits, mantissa_bits, exponent_bias)
+    dst2 = launch_upcast_emulated(dst2, exponent_bits, mantissa_bits, exponent_bias)
+
+    if not (torch.equal(dst, dst2)):
+
+        print('Error!!!')
+
+        dst = dst.cpu().detach().numpy()
+        dst2 = dst2.cpu().detach().numpy()
+        src = src.cpu().detach().numpy()
+
+        print(src[dst != dst2][0])
+        print(dst[dst != dst2][0])
+        print(dst2[dst != dst2][0])
+        print(hex(src.view(np.uint32)[dst != dst2][0]))
+        print(hex(dst.view(np.uint32)[dst != dst2][0]))
+        print(hex(dst2.view(np.uint32)[dst != dst2][0]))
+        print('')
+        raise ValueError('%d elements mismatch' % (dst != dst2).sum())
+
+
+def upcast_test(src_dtype, dst_dtype, exponent_bits, mantissa_bits, exponent_bias, max_repr):
+
+    numbits_src = exponent_bits + mantissa_bits + 1
+
+    src = launch_exhaustive_populate(src_dtype, 0, 65536, False, numbits_src, max_repr)
+
+    dst = launch_type_convert_triton(src, dst_dtype)
+    dst = launch_type_convert_triton(dst, torch.float32)
+
+    dst2 = launch_upcast_emulated(src, exponent_bits, mantissa_bits, exponent_bias)
+
+    assert(torch.equal(dst, dst2))
+
+
+
+torch.float8e4nv = torch.float8_e4m3fn
```

Comment:
ooc, why do we do this ?

Reply:
@zahimoud so that the name of the Torch dtype matches the Triton dtype, because we use `getattr` to obtain the dtype object from its name.